smart_card_connector_app/src/window-apps-displaying.js

```diff
@@ -0,0 +1,122 @@
+/** @license
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview This script implements displaying a list of connected
+ * applications in the Smart Card Connector App window.
+ */
+
+goog.provide('GoogleSmartCard.ConnectorApp.Window.AppsDisplaying');
+
+goog.require('GoogleSmartCard.DebugDump');
+goog.require('GoogleSmartCard.Logging');
+goog.require('GoogleSmartCard.PcscLiteServerClientsManagement.PermissionsChecking.KnownApp');
+goog.require('GoogleSmartCard.PcscLiteServerClientsManagement.PermissionsChecking.KnownAppsRegistry');
+goog.require('goog.array');
+goog.require('goog.asserts');
+goog.require('goog.dom');
+goog.require('goog.events.EventType');
+goog.require('goog.log.Logger');
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/** @const */
+var PermissionsChecking = GSC.PcscLiteServerClientsManagement.PermissionsChecking;
+
+/** @const */
+var KnownApp = PermissionsChecking.KnownApp;
+
+/**
+ * @type {!goog.log.Logger}
+ * @const
+ */
+var logger = GSC.Logging.getScopedLogger('ConnectorApp.MainWindow');
+
+/**
+ * @type {Element}
+ * @const
+ */
+var appListElement = goog.dom.getElement('app-list');
+
+/**
+ * @const
+ */
+var knownAppsRegistry = new PermissionsChecking.KnownAppsRegistry();
+
+/**
+ * @type {goog.Promise.<!Array.<KnownApp>>}
+ */
+var lastKnownAppsPromise = null;
+
+/**
+ * @param {!goog.Promise.<!Array.<KnownApp>>} knownAppsPromise
+ * @param {!Array.<string>} appList
+ * @param {Array.<KnownApp>} knownApps
+ * @return {boolean}
+ */
+var updateAppView = function(knownAppsPromise, appList, knownApps) {
+    if (knownAppsPromise !== lastKnownAppsPromise) return false;
+
+    GSC.Logging.checkWithLogger(logger, !goog.isNull(appListElement));
+    goog.asserts.assert(appListElement);
+
+    goog.dom.removeChildren(appListElement);
+
+    for (var i = 0; i < appList.length; i++) {
+      var text = knownApps && knownApps[i] ?
+                 knownApps[i].name :
+                 '<' + appList[i] + '>';
+      var newElement = goog.dom.createDom('li', undefined, text);
+      goog.dom.append(appListElement, newElement);
+    }
+
+    return true;
+}
+
+/**
+ * @param {!Array.<string>} appList
+ */
+var onUpdateListener = function(appList) {
+  logger.fine('Application list updated, refreshing the view. ' +
+              'New list of id\'s: ' + appList);
```

Comment:
It's better to use DebugDump module for formatting the array, because JavaScript default array-to-string conversion rules are quite weird:
`[].toString()==''
['a,b','c'].toString()=='a,b,c'`